server/service/listing.js

```diff
@@ -0,0 +1,182 @@
+        u.scoreParams.quality = scores[u.StoreId];
+    })
+}
+
+function sortBuckets(listingsByBucket){
+    return Object.keys(listingsByBucket)
+        .map(bucketId => {
+            if(bucketId == NUMBER_OF_BUCKETS){
+                console.log("Final Bucket - Sorting by distance");
+                return {bucketId: bucketId, listings: flattenListingsByQuality(sortStoresInBucket(listingsByBucket[bucketId], sortByDistance))};
+            }else{
+                return {bucketId: bucketId, listings: flattenListingsByQuality(sortStoresInBucket(listingsByBucket[bucketId], sortByRecency))};
+            }
+        })
+}
+
+function sortStoresInBucket(listingsByQuality, fn){
+    return Object.keys(listingsByQuality)
+        .map(quality => {return {
+            quality: quality,
+            listings: listingsByQuality[quality].sort(fn)
+        }})
+}
+
+function flattenListingsByQuality(listings){
+    return listings.sort((q1, q2) => {
+        if(q1.quality >= q2.quality){
+            return -1;
+        }
+        return 1;
+    }).flatMap(q => q.listings)
+}
+
+const sortByRecency = (u1, u2) => {
+    if(u1.Timestamp == u2.Timestamp){
+        return 0;
+    }
+    if(u1.Timestamp > u2.Timestamp){
+        return -1;
+    }
+    return 1;
+};
+
+
+
+const sortByDistance = (u1, u2) => {
+    if(u1.scoreParams.distance >= u2.scoreParams.distance){
+        return 1;
+    }
+    return -1;
+};
+
+
+function getBucket(storeUpdate){
+    let distVal = storeUpdate.scoreParams.distance / 1000.0;
+    if(distVal < 1){
+        return 1;
+    }
+    if(distVal < 3){
+        return 2;
+    }
+    if(distVal < 5){
+        return 3;
+    }
+    if(distVal < 10){
+        return 4;
+    }
+    return 5;
+}
+
+/**
+ * Returns scoring parameters for sorting listing
+ * @param storeUpdate - instance of a store info updates
+ * @param params - user input
+ * @returns {{quality, distance}} - quality rating for store data and distance from user
+ */
+function getScoringParameters(storeUpdate, params){
+    return {
+        quality: getQuality(storeUpdate), distance: getDistance(storeUpdate, params)
+    };
+}
+
+/**
+ * Finds crow-fly distance between store location and user location
+ * @param storeUpdate - instance of a store info
+ * @param params - user input
+ */
+function getDistance(storeUpdate, params){
+    return geolib.getDistance(
+        {latitude: params.location.lat, longitude: params.location.lng},
+        {latitude: storeUpdate.Latitude, longitude: storeUpdate.Longitude}
+    )
+}
+
+/**
+ * Returns 2 for High, 1 for Medium and 0 for Low
+ * @param storeUpdate  Store Update to be scored on quality
+ */
+function getQuality(storeUpdate){
+    let availabilityScore = getQualityAvailabilityInfo(storeUpdate);
+    let safetyScore = getQualitySafetyInfo(storeUpdate);
+    if(availabilityScore > safetyScore){
+        return availabilityScore;
+    }
+    return safetyScore;
+}
+
+function getQualityAvailabilityInfo(storeUpdate){
```

Comment:
would be nice if `getQualityAvailabilityInfo` and `getQualitySafetyInfo` could be replaced by 1 method that checks for the string length etc and returns the same ranks (since they are both identical methods)

Reply:
Right now they are. Keeping methods separate let me evolve separately